src/groups/bmq/bmqp/bmqp_schemalearner.cpp

```diff
@@ -129,31 +138,52 @@ SchemaLearner::Context SchemaLearner::createContext(int foreignId)
     return context->second;
 }
 
-void SchemaLearner::observe(Context&                     context,
-                            const MessagePropertiesInfo& input)
+SchemaLearner::SchemaPtr
+SchemaLearner::learn(Context&                     context,
+                     const MessagePropertiesInfo& input,
+                     const bdlbb::Blob&           blob)
 {
-    SchemaIdType inputId = input.schemaId();
+    const SchemaIdType inputId = input.schemaId();
 
     if (!isPresentAndValid(inputId)) {
         // Nothing to do
-        return;  // RETURN
+        return SchemaPtr();  // RETURN
     }
 
     BSLS_ASSERT_SAFE(context);
 
-    typedef bsl::pair<HandlesMap::iterator, bool> InsertResult;
-
     // Lookup the schema within this source
-    InsertResult lookupOrInsert = context->d_handles.emplace(inputId,
-                                                             HandlePtr());
-    HandlePtr&   contextHandle  = lookupOrInsert.first->second;
+    HandlesMap::iterator it = context->d_handles.find(inputId);
 
-    if (contextHandle) {
-        if (input.isRecycled()) {
-            contextHandle->d_schema_sp.reset();
-            // Must destroy previously learned Schema
+    if (it == context->d_handles.end()) {
+        HandlePtr newHandle(new (*d_allocator_p) SchemaHandle(inputId),
+                            d_allocator_p);
+        it = context->d_handles.emplace(inputId, newHandle, d_allocator_p)
+                 .first;
+    }
+
+    const HandlePtr& contextHandle = it->second;
+
+    BSLS_ASSERT_SAFE(contextHandle);
+
+    if (input.isRecycled()) {
+        contextHandle->d_schema_sp.reset();
+        // Must destroy previously learned Schema
+    }
+
+    const SchemaPtr& result = contextHandle->d_schema_sp;
+
+    if (!result) {
+        // Learn new Schema by reading all MessageProperties.
+        MessageProperties mps(d_allocator_p);
+        int               rc = mps.streamIn(blob, input, result);
+
+        if (rc == 0) {
```

Comment:
If `rc != 0` we will silently return an empty `SchemaPtr`. Should we log something or skip the message?